ale/transformation.py

```diff
@@ -279,41 +280,51 @@ def generate_rotations(self, frames, times, time_bias, rotation_type, sensor_fra
         quats_and_avs_per_frame = []
 
         logger.debug(f"Generate rotation times: {times}")
-        with ThreadPoolExecutor(max_workers=3) as executor:
+        with ThreadPoolExecutor(max_workers=1) as executor:
             futures = []
-
             for s, d in frames:
                 if exact_ck_times and len(times) > 1:
-                    function_args = {"startEt": start_et+time_bias, "stopEt": stop_et+time_bias, "toFrame": d, "refFrame": s, "mission": mission, "searchKernels": self.search_kernels, "ckQualities": ["smithed", "reconstructed"], "fullKernelPath": False}
+                    function_args = {"startEt": start_et+time_bias, "stopEt": stop_et+time_bias, "toFrame": d, "refFrame": s, "mission": mission, "searchKernels": self.search_kernels, "fullKernelPath": False}
+                    logger.debug(f"Exact CK times function args: {function_args}")
                     futures.append(executor.submit(spiceql_call, "getExactTargetOrientations", function_args, self.use_web))
                 else:
-                    function_args = {"toFrame": d, "refFrame": s, "mission": mission, "searchKernels": self.search_kernels}
+                    function_args = {"ets": times, "toFrame": d, "refFrame": s, "mission": mission, "searchKernels": self.search_kernels}
+                    logger.debug(f"Non-exact CK times function args: {function_args}")
                     futures.append(executor.submit(get_ephem_data, times, "getTargetOrientations", 150, self.use_web, function_args))
 
-            quats_and_avs_per_frame = np.array([future.result() for future in futures])
-        logger.debug(f"Quats and AVs per frame: {quats_and_avs_per_frame}")
+            quats_and_avs_per_frame = [future.result() for future in futures]
+            logger.debug(f"Returned Quats and AVs per frame: {quats_and_avs_per_frame}")
+            # quats_and_avs_per_frame = np.asarray(quats_and_avs_per_frame)
 
         if exact_ck_times and len(times) > 1 and len(frames) > 0:
             times = [quat_av[0] for quat_av in quats_and_avs_per_frame[0]]
-            quats_and_avs_per_frame = np.array([quats_and_avs[:,1:] for quats_and_avs in quats_and_avs_per_frame])
+            quats_and_avs_per_frame = [quats_and_avs[:][1:] for quats_and_avs in quats_and_avs_per_frame]
 
         for i, frame in enumerate(frames):
             quats_and_avs = quats_and_avs_per_frame[i]
+            logger.debug(f"Quats and AVs per frame: {quats_and_avs}")
             quats = np.zeros((len(times), 4))
             avs = []
-            _quats = np.array(quats_and_avs)[:, 0:4]
+            _quats = np.asarray(quats_and_avs)[:, 0:4]
             logger.debug(f"Quats: {_quats}")
             for j, quat in enumerate(_quats):
-                quats[j,:3] = quat[1:]
-                quats[j, 3] = quat[0]
+                quats[j, 0] = quat[3]
+                quats[j, 1:4] = quat[0:3]
```

Comment:
I don't know if this is right. Does this change how the elements in each quaternion are arranged? If the orientation calls aren't consistent with how the quaternions are returned that is an issue we need to fix

Reply:
yeah, I was wondering this too, it seems to be changing the order of the quaternions. I think this is more a mismatch of how SPICE returns quats vs what ALE/CSM wants.